### PR TITLE
Visualization tool

### PR DESCRIPTION
**Pull request by lzBryan**
Two tools included:
1) Diff tool to compare on-chain state with the configuration
2) Tool to visualize assets

* Added a state.ts to each sdk package to get the on-chain state of the contracts.
* Created visualization tasks for each contract type which reads the config, then get the state and diffs them
* Added a local cache for state to avoid having to restart the whole process each time a RPC fails
* Can use --diff to print only if there is a difference, and can use --full-output to toggle between full object output or just the difference.

Example commands:
`make -s compare-testnet VISUALIZE_ARGS_COMMON="--diff"`

`make -s print-token-testnet`

### CODE REVIEW

File: packages/stg-evm-v2/devtools/src/credit-messaging/config.ts [hunk 1]
```diff
@@ -17,6 +17,10 @@ export const configurePlanner: CreditMessagingConfigurator = createConfigureNode
     const planner = await sdk.getPlanner()
     if (areBytes32Equal(planner, config.planner)) return []
 
+    if (!config.planner) {
+        return [] // No planner to set, do not overwrite the existing one
+    }
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/config.ts:22]: This line should be placed on the very top so that no logic is executed if the value is nullish.

For the comparison, use `config.planner == null`

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    const edgePromises: Promise<OmniEdge<CreditMessagingEdgeConfig>>[] = []
+    for (const { vector, config } of configGraph.connections) {
+        edgePromises.push(
+            getFromCacheOrChain('CreditMessagingEdge', vector.from, () => getEdgeFromChain(vector, config)).then(
+                (edge) => {
+                    // Adapt to account for printing bigints
+                    if (edge.config.gasLimit) {
+                        edge.config.gasLimit = BigInt(edge.config.gasLimit)
+                    }
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:33]: Why is this necessary? Both the config and the SDK return value should be normalized by schemas to `bigint` so this will only ever be a `bigint`.

If not, this will not work for `0` values as `if (edge.config.gasLimit)` will skip to `else` if the value is `0`

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    point: OmniPoint,
+    existingConfig: CreditMessagingNodeConfig
+): Promise<OmniNode<CreditMessagingNodeConfig>> {
+    const logger = createModuleLogger(`State Loader for CreditMessaging`)
+    const createSdk = createCreditMessagingFactory(createConnectedContractFactory())
+    const sdk = await createSdk(point)
+
+    logger.debug(`Querying ${getNetworkNameForEid(point.eid)}...`)
+
+    const onChainAssets: MessagingAssetConfig = !existingConfig.assets
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:59]: Pls use `== null` for asserting nullish values

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                          await sdk.getAsset(assetId),
+                          assetId,
+                      ])
+                  )
+              ).filter(([assetAddress, _assetId]) => assetAddress)
+          )
+
+    const config: CreditMessagingNodeConfig = {
+        assets: onChainAssets,
+        planner: (await sdk.getPlanner()) ?? '0x',
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:74]: `0x` is not a valid address value. Plus why would this need a default?

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        maxAssetId: (await sdk.getMaxAssetId()) ?? 0,
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:75]: Why the default here?

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    const [sendLibrary, recvLibrary] = await Promise.all([
+        endpointSdk.getSendLibrary(vector.from.address, vector.to.eid),
+        endpointSdk.getReceiveLibrary(vector.from.address, vector.to.eid),
+    ])
+
+    const [executorConfig, sendUlnConfig] = sendLibrary
+        ? await Promise.all([
+              endpointSdk.getExecutorConfig(vector.from.address, sendLibrary ?? '0x', vector.to.eid),
+              endpointSdk.getUlnConfig(vector.from.address, sendLibrary ?? '0x', vector.to.eid),
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:102]: `0x` is not a valid address value

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+          ])
+        : [undefined, undefined]
+
+    const recvUlnConfig = recvLibrary
+        ? await endpointSdk.getUlnConfig(vector.from.address, recvLibrary[0] ?? '0x', vector.to.eid)
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:107]: `0x` is not a valid address value. Also if you want the `recvLibrary`'s address only, you can do:

```typescript
const [sendLibrary, [recvLibrary]] = await Promise.all([
      endpointSdk.getSendLibrary(vector.from.address, vector.to.eid),
      endpointSdk.getReceiveLibrary(vector.from.address, vector.to.eid),
])
```

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        : { requiredDVNs: [] }
+
+    const gasLimit = await sdk.getGasLimit(vector.to.eid)
+
+    const enforcedOptions: OAppEnforcedOption[] = []
+    const config: CreditMessagingEdgeConfig = {
+        enforcedOptions: enforcedOptions,
+        gasLimit: gasLimit ?? 0n,
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:115]: Why the default here?

File: packages/stg-evm-v2/devtools/src/credit-messaging/state.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        sendConfig: {
+            executorConfig: executorConfig,
+            ulnConfig: sendUlnConfig,
+        },
+        receiveConfig: { ulnConfig: recvUlnConfig },
+    }
+
+    if (!existingConfig.enforcedOptions) {
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/credit-messaging/state.ts:123]: Pls use `== null`

File: packages/stg-evm-v2/devtools/src/feeLib_v1/state.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+async function getNodeFromChain(point: OmniPoint): Promise<OmniNode<FeeLibV1NodeConfig>> {
+    const logger = createModuleLogger(`State Loader for FeeLibV1`)
+    const createSdk = createFeeLibV1Factory(createConnectedContractFactory())
+    const sdk = await createSdk(point)
+
+    logger.debug(`Querying ${getNetworkNameForEid(point.eid)}...`)
+
+    const config: FeeLibV1NodeConfig = { owner: (await sdk.getOwner()) ?? '0x' }
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/feeLib_v1/state.ts:30]: `0x` is not a valid address value. Also why the default here?

File: packages/stg-evm-v2/devtools/src/mintable/state.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    point: OmniPoint,
+    existingConfig: MintableNodeConfig
+): Promise<OmniNode<MintableNodeConfig>> {
+    const logger = createModuleLogger(`State Loader for Mintable`)
+    const createSdk = createMintableFactory(createConnectedContractFactory())
+    const sdk = await createSdk(point)
+
+    logger.debug(`Querying ${getNetworkNameForEid(point.eid)}...`)
+
+    if (!existingConfig.minters) {
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/mintable/state.ts:33]: `== null`

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    point: OmniPoint,
+    existingConfig: TokenMessagingNodeConfig
+): Promise<OmniNode<TokenMessagingNodeConfig>> {
+    const logger = createModuleLogger(`State Loader for TokenMessaging`)
+    const createSdk = createTokenMessagingFactory(createConnectedContractFactory())
+    const sdk = await createSdk(point)
+
+    logger.debug(`Querying ${getNetworkNameForEid(point.eid)}...`)
+
+    const onChainAssets: MessagingAssetConfig = !existingConfig.assets
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/token-messaging/state.ts:62]: `== null`

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                      Object.values(existingConfig.assets).map(async (assetId) => [
+                          await sdk.getAsset(assetId),
+                          assetId,
+                      ])
+                  )
+              ).filter(([assetAddress, _assetId]) => assetAddress)
+          )
+
+    const config: TokenMessagingNodeConfig = {
+        planner: (await sdk.getPlanner()) ?? '0x',
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/token-messaging/state.ts:76]: `0x` & no defaults

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        maxAssetId: (await sdk.getMaxAssetId()) ?? 0,
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/token-messaging/state.ts:77]: The same, no defaults

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+    const [sendLibrary, recvLibrary] = await Promise.all([
+        endpointSdk.getSendLibrary(vector.from.address, vector.to.eid),
+        endpointSdk.getReceiveLibrary(vector.from.address, vector.to.eid),
+    ])
+
+    const [executorConfig, sendUlnConfig] = sendLibrary
+        ? await Promise.all([
+              endpointSdk.getExecutorConfig(vector.from.address, sendLibrary ?? '0x', vector.to.eid),
+              endpointSdk.getUlnConfig(vector.from.address, sendLibrary ?? '0x', vector.to.eid),
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/token-messaging/state.ts:104]: `0x`

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        sdk.getGasLimit(vector.to.eid),
+        sdk.getNativeDropAmount(vector.to.eid),
+    ])
+
+    const config: TokenMessagingEdgeConfig = {
+        // Fares not part of the config, they are set by the planner
+        enforcedOptions: [],
+        maxPassengers: maxPassengers ?? 0,
+        gasLimit: gasLimit ?? { gasLimit: 0n, nativeDropGasLimit: 0n },
+        nativeDropAmount: nativeDropAmount ?? 0n,
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/token-messaging/state.ts:123]: Defaults again

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+    const enforcedOptions: OAppEnforcedOption[] = []
+    for (const enforcedOptionsConfig of existingConfig.enforcedOptions) {
+        const optionsHex = await oappSdk.getEnforcedOptions(vector.to.eid, enforcedOptionsConfig.msgType)
+        if (isZero(optionsHex)) continue
+
+        const options = Options.fromOptions(optionsHex)
+        enforcedOptions.push(...decodeEnforcedOptions(options, enforcedOptionsConfig.msgType))
+    }
+    config.enforcedOptions = enforcedOptions
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/token-messaging/state.ts:144]: This you could make more functional & parallelized by doing smth like:

```typescript
const enforcedOptions = await Promise.all(existingConfig.enforcedOptions.map(async () => {
    const optionsHex = await oappSdk.getEnforcedOptions(vector.to.eid, enforcedOptionsConfig.msgType)
    if (isZero(optionsHex)) return []

    const options = Options.fromOptions(optionsHex)
    return decodeEnforcedOptions(options, enforcedOptionsConfig.msgType)
})

config.enforcedOptions = enforcedOptions.flat()
```

File: packages/stg-evm-v2/devtools/src/treasurer/state.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    const onChainAssets = Object.fromEntries(
+        await Promise.all(
+            Object.keys(existingConfig.assets).map(async (asset) => {
+                return [asset, await sdk.getAsset(asset)]
+            })
+        )
+    )
+
+    const config: TreasurerNodeConfig = {
+        admin: (await sdk.getAdmin()) ?? '0x',
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/src/treasurer/state.ts:42]: `0x`

File: packages/stg-evm-v2/devtools/tasks/visualize.ts [hunk 19]
```diff
@@ -0,0 +1,663 @@
...
+}
+
+function stripAssetId(configGraph: AssetOmniGraph): AssetOmniGraph {
+    for (const { config } of configGraph.contracts) {
+        delete config.assetId
+    }
+    return configGraph
+}
+
+// TODO: Fix this
```
Comment (janjakubnanista) [inline packages/stg-evm-v2/devtools/tasks/visualize.ts:492]: Yes please

**Review by janjakubnanista — CHANGES_REQUESTED** [review]
There's still a bunch of TODOs left, what's the plan there?

File: Makefile [hunk 20]
```diff
@@ -335,3 +335,45 @@ transfer-mainnet:
...
+	# $(HARDHAT) stg:visualize::feelib-v1 $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/feelib-v1.usdt.config.ts --contract-name FeeLibV1USDT
+	# $(HARDHAT) stg:visualize::asset $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/asset.eth.config.ts --contract-name Asset
+	# $(HARDHAT) stg:visualize::asset $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/asset.usdc.config.ts --contract-name Asset
+	# $(HARDHAT) stg:visualize::asset $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/asset.usdt.config.ts --contract-name Asset
+	# $(HARDHAT) stg:visualize::treasurer $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/treasurer.config.ts --contract-name Treasurer
+	# $(HARDHAT) stg:visualize::staking $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/staking.config.ts --contract-name Staking
+	# $(HARDHAT) stg:visualize::rewarder $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/rewarder.config.ts --contract-name Rewarder
+	# $(HARDHAT) stg:visualize::oft $(VISUALIZE_ARGS_COMMON) --oapp-config $(CONFIG_BASE_PATH)/oft-token.config.ts --contract-name OFT
```
[the remark this answers is not in the document]
Comment (lzBryan) [reply Makefile:360]: Whoops, no, will un-comment.

File: packages/stg-evm-v2/devtools/src/asset/sdk.ts [hunk 21]
```diff
@@ -79,6 +88,17 @@ export class Asset extends Ownable implements IAsset {
         }
     }
 
+    @AsyncRetriable()
+    async getStargateType(): Promise<StargateType> {
+        const stargateType = await this.contract.contract.stargateType()
+
+        if (stargateType === 0) {
+            return StargateType.Pool
+        }
```
[the remark this answers is not in the document]
Comment (lzBryan) [reply packages/stg-evm-v2/devtools/src/asset/sdk.ts:97]: Sounds good, will do.

File: packages/stg-evm-v2/devtools/src/credit-messaging/config.ts (hunk 1 above)
Comment (lzBryan) [reply packages/stg-evm-v2/devtools/src/credit-messaging/config.ts:22]: Sounds good, will do.

File: packages/stg-evm-v2/devtools/src/token-messaging/state.ts (hunk 17 above)
Comment (lzBryan) [reply packages/stg-evm-v2/devtools/src/token-messaging/state.ts:144]: Thank you!
